src/transformers/cache_utils.py

```diff
@@ -184,6 +187,120 @@ def from_legacy_cache(cls, past_key_values: Optional[Tuple[Tuple[torch.FloatTens
+            self.value_cache.append(torch.zeros(0, dtype=key_states.dtype, device=key_states.device))
+            keys_to_return, values_to_return = key_states, value_states
+        else:
+            dequant_key = self._key_cache_quant[layer_idx].dequantize()
+            dequant_value = self._value_cache_quant[layer_idx].dequantize()
+            keys_to_return = torch.cat(
+                [
+                    dequant_key,
+                    self.key_cache[layer_idx],
+                    key_states,
+                ],
+                dim=-2,
+            )
+            values_to_return = torch.cat(
+                [
+                    dequant_value,
+                    self.value_cache[layer_idx],
+                    value_states,
+                ],
+                dim=-2,
+            )
+            if (
+                self.key_cache[layer_idx].dim() == 4
+                and self.key_cache[layer_idx].shape[-2] + 1 == self.residual_length
+            ):
+                self._key_cache_quant[layer_idx] = self.quantize(keys_to_return.contiguous())
+                self._value_cache_quant[layer_idx] = self.quantize(values_to_return.contiguous())
+                self.key_cache[layer_idx] = torch.zeros(0, dtype=key_states.dtype, device=key_states.device)
+                self.value_cache[layer_idx] = torch.zeros(0, dtype=key_states.dtype, device=key_states.device)
+            else:
+                self.key_cache[layer_idx] = torch.cat([self.key_cache[layer_idx], key_states], dim=-2)
+                self.value_cache[layer_idx] = torch.cat([self.value_cache[layer_idx], value_states], dim=-2)
+
+        return keys_to_return, values_to_return
+
+    def get_seq_length(self, layer_idx: Optional[int] = 0) -> int:
+        """Returns the sequence length of the cached states. A layer index can be optionally passed."""
+        if len(self.key_cache) <= layer_idx:
+            return 0
+        return self.seen_token
+
+    def get_max_length(self) -> Optional[int]:
+        """Returns the maximum sequence length of the cached states. DynamicCache does not have a maximum length."""
+        return None
+
+    def reorder_cache(self, beam_idx: torch.LongTensor):
+        """Reorders the cache for beam search, given the selected beam indices."""
+        pass
+
+    def to_legacy_cache(self) -> Tuple[Tuple[torch.Tensor], Tuple[torch.Tensor]]:
+        """Converts the `DynamicCache` instance into the its equivalent in the legacy cache format."""
+        return None
+
+    @classmethod
+    def from_legacy_cache(cls, past_key_values: Optional[Tuple[Tuple[torch.FloatTensor]]] = None) -> "DynamicCache":
+        """Converts a cache in the legacy cache format into an equivalent `DynamicCache`."""
+        return None
+
+    def quantize(self, tensor):
+        qtensor = QBitsTensor.quantize(tensor, axis=0, qtype=self.qtype, group_size=self.q_group_size)
```

Comment:
Shouldn't the quantization happen across different dimensions for the key and value? According to the paper, the key has per-channel quantization (axis=3?) and the value has per-token quantization (axis=2?) 

Recall: the key and value tensors have shape `(batch_size, num_heads, sequence_length, embed_size_per_head)`

Reply:
Yes, the paper does it slightly differently but we do not follow the same recipe as they did. It was more like take their results and build on them.

I tried using per-tensor quantization actually and it gives almost same results in terms of speed and quality, so I opted for being consistent and easy to maintain by using one quantization method. I will add this in __init__ docs for Cache class